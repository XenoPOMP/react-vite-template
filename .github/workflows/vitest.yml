name: 'Testing'
on:
  push:
  pull_request:

jobs:
  test:
    runs-on: ubuntu-latest

    permissions:
      # Required to checkout the code
      contents: read
      # Required to put a comment into the pull-request
      pull-requests: write

    steps:
      - uses: actions/checkout@v2
      - name: 'Install Node'
        uses: actions/setup-node@v2
        with:
          node-version: '16.x'
      - name: 'Install Deps'
        run: npm i yarn -g && yarn
      - name: 'Test'
        run: yarn coverage
      - name: 'Report Coverage'
        if: always() # Also generate the report if tests are failing
        uses: davelosert/vitest-coverage-report-action@v2
