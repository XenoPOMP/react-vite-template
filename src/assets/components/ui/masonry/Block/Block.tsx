import { PropsWith } from '@xenopomp/advanced-types';

import cn from 'classnames';
import { FC } from 'react';

import styles from './Block.module.scss';
import type { BlockProps } from './Block.props';

/**
 * See docs in {@link Masonry}.
 *
 * @param className
 * @param style
 * @param id
 * @param children
 * @constructor
 */
const Block: FC<
  PropsWith<'className' | 'style' | 'id' | 'children', BlockProps>
> = ({ className, style, id, children }) => {
  return (
    <div className={cn(styles.block, className)} id={id} style={style}>
      {children}
    </div>
  );
};

export default Block;
