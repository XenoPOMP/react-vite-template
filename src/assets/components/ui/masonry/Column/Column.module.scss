.column {
  width: 100%;
  display: flex;
  justify-content: start;
  align-items: center;
  flex-direction: column;
  gap: inherit;
}
