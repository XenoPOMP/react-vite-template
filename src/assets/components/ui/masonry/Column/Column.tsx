import { PropsWith } from '@xenopomp/advanced-types';

import cn from 'classnames';
import { FC } from 'react';

import styles from './Column.module.scss';
import type { ColumnProps } from './Column.props';

/**
 * See docs in {@link Masonry}.
 *
 * @param children
 * @param className
 * @param style
 * @param id
 * @constructor
 */
const Column: FC<
  PropsWith<'className' | 'style' | 'id' | 'children', ColumnProps>
> = ({ children, className, style, id }) => {
  return (
    <div className={cn(styles.column, className)} id={id} style={style}>
      {children}
    </div>
  );
};

export default Column;
