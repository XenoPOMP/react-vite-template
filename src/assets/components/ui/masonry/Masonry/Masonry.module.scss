.masonry {
  display: grid;
  grid-template-columns: repeat(var(--column-count), 1fr);
  width: 100%;
  font-size: 1em;
  gap: .5em;
}
