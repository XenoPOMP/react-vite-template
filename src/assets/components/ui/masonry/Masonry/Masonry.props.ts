export interface MasonryProps {
  /** Defines masonry`s default column count. */
  columns?: number;
}
