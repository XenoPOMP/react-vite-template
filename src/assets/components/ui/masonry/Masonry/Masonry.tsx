import { PropsWith } from '@xenopomp/advanced-types';

import cn from 'classnames';
import { CSSProperties, FC } from 'react';

import styles from './Masonry.module.scss';
import type { MasonryProps } from './Masonry.props';

/**
 * Masonry grid component.
 *
 * @param columns
 * @param className
 * @param style
 * @param id
 * @param children
 * @constructor
 *
 * @example
 * <Masonry>
 *   <Column>
 *     <Block className={cn('border border-black h-[100px]')}></Block>
 *     <Block className={cn('border border-black h-[200px]')}></Block>
 *   </Column>
 *
 *   <Column>
 *     <Block className={cn('border border-black h-[200px]')}></Block>
 *     <Block className={cn('border border-black h-[100px]')}></Block>
 *   </Column>
 *
 *   <Column>
 *     <Block className={cn('border border-black h-[100px]')}></Block>
 *     <Block className={cn('border border-black h-[200px]')}></Block>
 *   </Column>
 * </Masonry>
 */
const Masonry: FC<
  PropsWith<'className' | 'style' | 'id' | 'children', MasonryProps>
> = ({ columns, className, style, id, children }) => {
  return (
    <div
      className={cn(styles.masonry, className)}
      style={
        { ...style, '--column-count': columns ? columns : 3 } as CSSProperties
      }
      id={id}
    >
      {children}
    </div>
  );
};

export default Masonry;
