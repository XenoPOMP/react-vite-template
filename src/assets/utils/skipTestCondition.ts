import useEnv, { TestingMode } from '@hooks/useEnv';

const { TESTING_MODE } = useEnv();

/**
 * Determines skip test condition.
 * @param {TestingMode} mode       target mode
 * @returns {boolean}              false if current mode isn`t the same as target one
 *                                 or current mode is FULLSTACK
 */
const skipTestCondition = (mode: TestingMode): boolean => {
  if (TESTING_MODE === 'FULLSTACK') {
    return false;
  }

  return TESTING_MODE !== mode;
};

export { skipTestCondition as default };
