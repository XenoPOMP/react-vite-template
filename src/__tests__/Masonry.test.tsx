import { render } from '@testing-library/react';
import { describe, expect, test } from 'vitest';

import Block from '@ui/masonry/Block/Block';
import Column from '@ui/masonry/Column/Column';
import Masonry from '@ui/masonry/Masonry/Masonry';

import skipTestCondition from '@utils/skipTestCondition';

describe.skipIf(skipTestCondition('FRONTEND'))('Masonry gallery', () => {
  test('Match snapshot', () => {
    expect(
      render(
        <Masonry>
          <Column>
            <Block>1</Block>
            <Block>2</Block>
            <Block>3</Block>
          </Column>

          <Column>
            <Block>4</Block>
            <Block>5</Block>
            <Block>6</Block>
          </Column>

          <Column>
            <Block>7</Block>
            <Block>8</Block>
            <Block>9</Block>
          </Column>
        </Masonry>
      )
    ).toMatchSnapshot();

    expect(
      render(
        <Masonry columns={4}>
          <Column>
            <Block>1</Block>
            <Block>2</Block>
            <Block>3</Block>
          </Column>

          <Column>
            <Block>4</Block>
            <Block>5</Block>
            <Block>6</Block>
          </Column>

          <Column>
            <Block>7</Block>
            <Block>8</Block>
            <Block>9</Block>
          </Column>

          <Column>
            <Block>7</Block>
            <Block>8</Block>
            <Block>9</Block>
          </Column>
        </Masonry>
      )
    ).toMatchSnapshot();
  });
});
