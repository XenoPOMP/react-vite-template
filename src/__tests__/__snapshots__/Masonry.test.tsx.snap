// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Masonry gallery > Match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_masonry_0a77d1"
        style="--column-count: 3;"
      >
        <div
          class="_column_225756"
        >
          <div
            class="_block_229b1d"
          >
            1
          </div>
          <div
            class="_block_229b1d"
          >
            2
          </div>
          <div
            class="_block_229b1d"
          >
            3
          </div>
        </div>
        <div
          class="_column_225756"
        >
          <div
            class="_block_229b1d"
          >
            4
          </div>
          <div
            class="_block_229b1d"
          >
            5
          </div>
          <div
            class="_block_229b1d"
          >
            6
          </div>
        </div>
        <div
          class="_column_225756"
        >
          <div
            class="_block_229b1d"
          >
            7
          </div>
          <div
            class="_block_229b1d"
          >
            8
          </div>
          <div
            class="_block_229b1d"
          >
            9
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_masonry_0a77d1"
      style="--column-count: 3;"
    >
      <div
        class="_column_225756"
      >
        <div
          class="_block_229b1d"
        >
          1
        </div>
        <div
          class="_block_229b1d"
        >
          2
        </div>
        <div
          class="_block_229b1d"
        >
          3
        </div>
      </div>
      <div
        class="_column_225756"
      >
        <div
          class="_block_229b1d"
        >
          4
        </div>
        <div
          class="_block_229b1d"
        >
          5
        </div>
        <div
          class="_block_229b1d"
        >
          6
        </div>
      </div>
      <div
        class="_column_225756"
      >
        <div
          class="_block_229b1d"
        >
          7
        </div>
        <div
          class="_block_229b1d"
        >
          8
        </div>
        <div
          class="_block_229b1d"
        >
          9
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Masonry gallery > Match snapshot 2`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_masonry_0a77d1"
        style="--column-count: 3;"
      >
        <div
          class="_column_225756"
        >
          <div
            class="_block_229b1d"
          >
            1
          </div>
          <div
            class="_block_229b1d"
          >
            2
          </div>
          <div
            class="_block_229b1d"
          >
            3
          </div>
        </div>
        <div
          class="_column_225756"
        >
          <div
            class="_block_229b1d"
          >
            4
          </div>
          <div
            class="_block_229b1d"
          >
            5
          </div>
          <div
            class="_block_229b1d"
          >
            6
          </div>
        </div>
        <div
          class="_column_225756"
        >
          <div
            class="_block_229b1d"
          >
            7
          </div>
          <div
            class="_block_229b1d"
          >
            8
          </div>
          <div
            class="_block_229b1d"
          >
            9
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        class="_masonry_0a77d1"
        style="--column-count: 4;"
      >
        <div
          class="_column_225756"
        >
          <div
            class="_block_229b1d"
          >
            1
          </div>
          <div
            class="_block_229b1d"
          >
            2
          </div>
          <div
            class="_block_229b1d"
          >
            3
          </div>
        </div>
        <div
          class="_column_225756"
        >
          <div
            class="_block_229b1d"
          >
            4
          </div>
          <div
            class="_block_229b1d"
          >
            5
          </div>
          <div
            class="_block_229b1d"
          >
            6
          </div>
        </div>
        <div
          class="_column_225756"
        >
          <div
            class="_block_229b1d"
          >
            7
          </div>
          <div
            class="_block_229b1d"
          >
            8
          </div>
          <div
            class="_block_229b1d"
          >
            9
          </div>
        </div>
        <div
          class="_column_225756"
        >
          <div
            class="_block_229b1d"
          >
            7
          </div>
          <div
            class="_block_229b1d"
          >
            8
          </div>
          <div
            class="_block_229b1d"
          >
            9
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_masonry_0a77d1"
      style="--column-count: 4;"
    >
      <div
        class="_column_225756"
      >
        <div
          class="_block_229b1d"
        >
          1
        </div>
        <div
          class="_block_229b1d"
        >
          2
        </div>
        <div
          class="_block_229b1d"
        >
          3
        </div>
      </div>
      <div
        class="_column_225756"
      >
        <div
          class="_block_229b1d"
        >
          4
        </div>
        <div
          class="_block_229b1d"
        >
          5
        </div>
        <div
          class="_block_229b1d"
        >
          6
        </div>
      </div>
      <div
        class="_column_225756"
      >
        <div
          class="_block_229b1d"
        >
          7
        </div>
        <div
          class="_block_229b1d"
        >
          8
        </div>
        <div
          class="_block_229b1d"
        >
          9
        </div>
      </div>
      <div
        class="_column_225756"
      >
        <div
          class="_block_229b1d"
        >
          7
        </div>
        <div
          class="_block_229b1d"
        >
          8
        </div>
        <div
          class="_block_229b1d"
        >
          9
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
